### DISCUSSION TITLE
Show file history

### DISCUSSION THREAD

**Discussion by aawsome** [rustic-rs/rustic#1238] [Show and tell]
I'd like to show some functionality which has been introduced in rustic 0.8.0.

Sometimes you know about a file which is present in your repository and you want to know when this exactly given file has been changed. Basically you want to see the history of this file in the snapshots saved in your repository.

To do so, simply use the `--path` option of `rustic find`. This will find the given `path` in *all* snapshots. However, by default identical results are shown as `(+x)` (like in the `rustic snapshots` output). So, this gives you exactly the history of that file.

Typically, you also know that you can filter out some snapshots by using the path(s) stored in the snapshots, so using a suitable `--filter-paths` will reduce the search by excluding snapshots which cannot contain that file. 

An example call:
```
$ rustic find --path /home/alex/.config/mc/ini --filter-paths /home
[00:00:01] reading index...               ████████████████████████████████████████        451/451

searching in snapshots group (host [myhost], label [], paths [/home])...
found in 03521ca4 from 2022-06-11 06:29:55
-rw-r--r--     alex     alex      4036 28 Mar 07:22 "/home/alex/.config/mc/ini" 
found in 772fc402 from 2022-06-30 23:30:21 (+1)
-rw-r--r--     alex     alex      4036 13 Jun 08:19 "/home/alex/.config/mc/ini" 
found in dd4d75e2 from 2022-08-31 23:30:33 (+16)
-rw-r--r--     alex     alex      4036 12 Aug 12:55 "/home/alex/.config/mc/ini" 
found in b31f134e from 2023-10-29 23:30:20 (+230)
-rw-r--r--     alex     alex      4078 24 Oct 06:41 "/home/alex/.config/mc/ini" 
found in bfacf88a from 2024-01-29 09:30:27 (+256)
-rw-r--r--     alex     alex      4078 28 Jan 23:32 "/home/alex/.config/mc/ini" 
found in fef232e8 from 2024-04-23 23:30:52 (+432)
-rw-r--r--     alex     alex      4078 15 Apr 12:18 "/home/alex/.config/mc/ini" 
```

This shows that the given file exists in quite some snapshots - we exactly get the snapshots where the file content changed.

**Comment by simonsan** [comment]
Would be nice, to add this to the docs actually, e.g. here: https://github.com/rustic-rs/docs/tree/main/src/commands/misc

As a use case for finding things.

Ref.: https://github.com/rustic-rs/docs/issues/83

**Reply by simonsan** [reply]
Online: https://rustic.cli.rs/docs/commands/misc/file_history.html

**Comment by nickchomey** [comment]
I don't understand what this is doing/showing. If you make a doc on it, can you please provide more detail on how to interpret and use the output?

I don't know what +x means, or what differs/remains the same between each result.

**Reply by aawsome** [reply]
@nickchomey This command searches for the given path in all snapshots and shows the finds. The snapshots are sorted by the snapshot timestamp. Moreover, by default identical nodes are grouped together, this means that 
```
found in dd4d75e2 from 2022-08-31 23:30:33 (+16)
-rw-r--r--     alex     alex      4036 12 Aug 12:55 "/home/alex/.config/mc/ini" 
```
means that this file was found in the shown snapshot `dd4d75e2` and in the 16 subsequent snapshots. Moreover, the contents and metadata of it is exactly the same for all 17 snapshots.

Note, this also works for dirs - if `rustic find` groups dirs which are founds, the content and metadata of the dir for all contained files and subdirs are exactly identical.

So, this grouping only shows that there is something different or same. To really see what is different, you have to use `rustic diff` or maybe even show the contents using `rustic dump`.

**Reply by nickchomey** [reply]
Thanks! So, with your original example

```
found in 03521ca4 from 2022-06-11 06:29:55
-rw-r--r--     alex     alex      4036 28 Mar 07:22 "/home/alex/.config/mc/ini" 
found in 772fc402 from 2022-06-30 23:30:21 (+1)
-rw-r--r--     alex     alex      4036 13 Jun 08:19 "/home/alex/.config/mc/ini" 
found in dd4d75e2 from 2022-08-31 23:30:33 (+16)
-rw-r--r--     alex     alex      4036 12 Aug 12:55 "/home/alex/.config/mc/ini" 
found in b31f134e from 2023-10-29 23:30:20 (+230)
-rw-r--r--     alex     alex      4078 24 Oct 06:41 "/home/alex/.config/mc/ini" 
found in bfacf88a from 2024-01-29 09:30:27 (+256)
-rw-r--r--     alex     alex      4078 28 Jan 23:32 "/home/alex/.config/mc/ini" 
found in fef232e8 from 2024-04-23 23:30:52 (+432)
-rw-r--r--     alex     alex      4078 15 Apr 12:18 "/home/alex/.config/mc/ini"
```

this means that the file was 
* created on 28 Mar (of which year? it doesnt show, presumably because `ls` doesnt show it by default. Perhaps a standard dd/mm/yyyy can be used in both places. eg. `ls -l --time-style=+%d/%m/%Y?)`
* added to the repo on 2022-06-11
* changed on 13 Jun and snapshotted 30 Jun, didnt change for the next 1 snapshot 
* changed 12 Aug, snapshotted 2022-08-31 and it didnt change for 16 snapshots
* etc...

Is that correct?

**Reply by aawsome** [reply]
exactly! Thanks for the hint of the missing year; I opened https://github.com/rustic-rs/rustic/issues/1239

**Reply by aawsome** [reply]
Just a note: Of course there could have be more changes of this file in-between snapshots. But rustic can only detect changes between the snapshot times..

**Reply by nickchomey** [reply]
Great, thanks for clarifying/confirming! 

And, yes, of course there could have been uncaptured changes between snapshots. We'd need some real-time file watcher sync thing (eg like onedrive does on windows) with versioning. I doubt that's in the scope of rustic.

**Reply by simonsan** [reply]
> Great, thanks for clarifying/confirming!
> 
> And, yes, of course there could have been uncaptured changes between snapshots. We'd need some real-time file watcher sync thing (eg like onedrive does on windows) with versioning. I doubt that's in the scope of rustic.

You could do that with a script and https://github.com/watchexec/watchexec/ for example
